src/Type/Set/VariantF.hs

```diff
@@ -0,0 +1,194 @@
+  VariantF :: ()
+    => SSide ss -> Follow ss bst a -> VariantF bst a
+type role VariantF nominal nominal
+
+
+------------------------------------------------------------------------------
+-- | A proof that the set @bst@ contains functor @f@.
+class HasF f bst where
+
+  -- | Inject a @t@ into a 'VariantF'.
+  toVariantF :: f a -> VariantF bst a
+
+  -- | Attempt to project a 'VariantF' into @f a@. This might fail, because there
+  -- is no guarantee that the 'VariantF' /actually contains/ @f a@.
+  fromVariantF :: VariantF bst a -> Maybe (f a)
+
+instance ( Follow (Locate f bst) bst ~ f
+         , FromSides (Locate f bst)
+         ) => HasF f bst where
+  toVariantF = VariantF (fromSides @(Locate f bst))
+  fromVariantF (VariantF tag res) =
+    case testEquality tag (fromSides @(Locate f bst)) of
+      Just Refl -> Just res
+      Nothing -> Nothing
+
+-- forall c f bst. (ForAll c bst, HasF f bst) => c f
+
+------------------------------------------------------------------------------
+-- | TODO
+data SplitF f lbst rbst a
+  = RootF (f a)
+  | LSplitF (VariantF lbst a)
+  | RSplitF (VariantF rbst a)
+
+------------------------------------------------------------------------------
+-- | TODO
+decompRootF :: VariantF ('Branch f lbst rbst) a -> SplitF f lbst rbst a
+decompRootF (VariantF SNil   t) = RootF t
+decompRootF (VariantF (SL s) t) = LSplitF (VariantF s t)
+decompRootF (VariantF (SR s) t) = RSplitF (VariantF s t)
+
+------------------------------------------------------------------------------
+-- | A proof that inserting into a @bst@ doesn't affect the position of
+-- anything already in the tree.
+proveFollowInsertF :: forall (ss :: [Side]) (f :: * -> *) (bst :: TypeSet (* -> *)).
+  Follow ss (Insert f bst) :~~: Follow ss bst
+proveFollowInsertF = unsafeCoerce HRefl
+
+
+
+------------------------------------------------------------------------------
+-- | Weaken a 'VariantF' so that it can contain something else.
+weakenF :: forall f bst a. VariantF bst a -> VariantF (Insert f bst) a
+weakenF (VariantF (tag :: SSide ss) res)
+  = VariantF (tag :: SSide ss) $
+    case proveFollowInsertF @ss @f @bst of
+      HRefl -> res :: Follow ss bst a
+
+
+instance (ForAllIn Functor bst) => Functor (VariantF bst) where
+  fmap f (VariantF s r) = case forMember @_ @Functor @bst s of
+    Dict -> VariantF s $ fmap f r
+
+instance ( ForAllIn Functor bst
+         , ForAllIn Foldable bst
+         ) => Foldable (VariantF bst) where
+  foldMap f (VariantF s r)
+    = case forMember @_ @Foldable @bst s of
+        Dict -> foldMap f r
+
+instance ( ForAllIn Functor bst
+         , ForAllIn Foldable bst
+         , ForAllIn Traversable bst
+         ) => Traversable (VariantF bst) where
+  traverse f (VariantF s r)
+    = case forMember @_ @Traversable @bst s of
+        Dict -> VariantF s <$> traverse f r
+
+data SomeType = SomeType
+
+-- | I have no good way to ensure that a is typeable so the output replaces
```

Comment:
What about `instance (Typeable a, ForAllIn (forall x. Show x => Show (f x)) bst) => Show (VariantF bst)`, where you somehow bind `f` in the `ForAllIn`?

Reply:
If I were just going to skip the `Show1` instance for `Show` I'd just go with `instance (Typeable a, ForAllIn Typeable bst, Show a, ForAllIn Show1 bst) => Show (VariantF bst a)`